react/explainer.md

```diff
@@ -0,0 +1,217 @@
+# Using React/Preact to implement AMP elements
+
+Discussion: [issues](https://github.com/ampproject/wg-bento/labels/Discussion%3A%20React), [new issue](https://github.com/ampproject/wg-bento/issues/new?labels=Discussion%3A+React).
+
+Prototype: [repo](https://github.com/ampproject/amp-react-prototype), [decision points](https://github.com/ampproject/amp-react-prototype/labels/TBD).
+
+## Concept
+
+The concept of building AMP extensions using React/Preact has been introduced
+in the [AMP Contributor Summit 2019 talk](https://www.youtube.com/watch?v=s78VcduTOqE).
+
+With Bento, we will need to support the following component modes:
+ 1. AMP mode: run on an AMP page;
+ 2. "Bento" mode: run on a regular web page not managed by the AMP framework;
+ 3. 3P framework mode: allowing a 3P framework to directly use an AMP component.
+
+We are making the AMP document framework more modular and configurable. But we
+also need to make components more indepedent, mutable, and distributable.
+
+The AMP elements have the following needs:
+ 1. We need some internal state about the component (which carousel slide am I on?)
+ 2. We need external context to tell the component when it needs to perform
+    resource-intensive actions.
+ 3. We need a way to render this in DOM.
+ 4. We need to react to component mutations and state changes.
+
+Mutability is the key requirement for Bento. Currently, if an AMP element needs
+to support an attribute or a child mutation, it has to manually add a listener
+or a mutation observer and apply changes. This architecture is difficult to
+maintain and error-prone. Instead we'd like AMP elements to be mutable "by design".
+
+To simplify the mutability problem we need the main part of the AMP element's
+implementation to be reenterable. What if we just had a "render" method?
+ - State changed? Call the render method.
+ - Framework says it’s ok to make requests now? Call the render method.
+ - Element resized? Call the render method.
+ - Element is visible in user’s viewport? Call the render method.
+
+React model fits this picture very well: there's a single "render" method and
+rendering side-effects with internal state.
```

Comment:
I would add: "This completely isolates the UI and rendering from the external environment, meaning we can reuse the same code in each of the 3 Modes."

Reply:
Done